dulwich/objects.py

```diff
@@ -135,7 +135,7 @@ def hex_to_filename(path, hex):
     # os.path.join accepts bytes or unicode, but all args must be of the same
     # type. Make sure that hex which is expected to be bytes, is the same type
     # as path.
-    if getattr(path, "encode", None) is not None:
+    if type(path) != type(hex) and getattr(path, "encode", None) is not None:
```

Comment:
This looks like just a typo to me; shouldn't it be checking for "decode" rather than "encode" here, in which case we don't need the type check?